src/util/simplify_expr_pointer.cpp

```diff
@@ -417,39 +418,67 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
   return true;
 }
 
+static exprt simplify_inequality_address_of_prepare_op(const exprt &op)
+{
+  exprt result = op;
+  while(result.id() == ID_typecast)
+  {
+    result = result.op0();
+  }
+  POSTCONDITION(result.id() == ID_address_of);
+  return result;
+}
+
+// returns true if op is a symbol or an expression like x.y[constant].z
+static bool simplify_inequality_is_simple_structure_expression(const exprt &op)
+{
+  exprt expr = op;
+  while(expr.id() == ID_member || expr.id() == ID_index)
+  {
+    if(expr.id() == ID_member)
+    {
+      expr = to_member_expr(expr).compound();
+    }
+    if(expr.id() == ID_index)
+    {
+      const index_exprt index = to_index_expr(expr);
+      if(!index.index().is_constant())
+      {
+        return false;
+      }
+      expr = index.array();
+    }
+  }
+  return expr.id() == ID_symbol;
+}
+
 bool simplify_exprt::simplify_inequality_address_of(exprt &expr)
 {
-  assert(expr.type().id()==ID_bool);
-  assert(expr.operands().size()==2);
-  assert(expr.id()==ID_equal || expr.id()==ID_notequal);
+  PRECONDITION(expr.type().id() == ID_bool);
+  PRECONDITION(expr.operands().size() == 2);
+  PRECONDITION(expr.id() == ID_equal || expr.id() == ID_notequal);
 
-  exprt tmp0=expr.op0();
-  if(tmp0.id()==ID_typecast)
-    tmp0=expr.op0().op0();
-  if(tmp0.op0().id()==ID_index &&
-     to_index_expr(tmp0.op0()).index().is_zero())
-    tmp0=address_of_exprt(to_index_expr(tmp0.op0()).array());
-  exprt tmp1=expr.op1();
-  if(tmp1.id()==ID_typecast)
-    tmp1=expr.op1().op0();
-  if(tmp1.op0().id()==ID_index &&
-     to_index_expr(tmp1.op0()).index().is_zero())
-    tmp1=address_of_exprt(to_index_expr(tmp1.op0()).array());
-  assert(tmp0.id()==ID_address_of);
-  assert(tmp1.id()==ID_address_of);
-
-  if(tmp0.operands().size()!=1)
-    return true;
-  if(tmp1.operands().size()!=1)
+  exprt lhs = simplify_inequality_address_of_prepare_op(expr.op0());
+  exprt rhs = simplify_inequality_address_of_prepare_op(expr.op1());
+
+  if(lhs.operands().size() != 1 || rhs.operands().size() != 1)
     return true;
 
-  if(tmp0.op0().id()==ID_symbol &&
-     tmp1.op0().id()==ID_symbol)
-  {
-    bool equal=
-       tmp0.op0().get(ID_identifier)==
-       tmp1.op0().get(ID_identifier);
+  // strip the address-of operator
+  lhs = lhs.op0();
+  rhs = rhs.op0();
 
+  if(
+    simplify_inequality_is_simple_structure_expression(lhs) &&
+    simplify_inequality_is_simple_structure_expression(rhs))
+  {
+    object_descriptor_exprt lhs_descriptor, rhs_descriptor;
+    lhs_descriptor.build(lhs, ns);
+    rhs_descriptor.build(rhs, ns);
+    exprt lhs_offset = simplify_expr(lhs_descriptor.offset(), ns);
+    exprt rhs_offset = simplify_expr(rhs_descriptor.offset(), ns);
+    bool equal = lhs_descriptor.root_object() == rhs_descriptor.root_object() &&
+                 lhs_offset.is_constant() && lhs_offset == rhs_offset;
```

Comment:
This looks dodgy since sometimes it really indicates we don't know if they're equal, but I think by `is_simple_structure_expression` above we always have `lhs_offset.is_constant()`? If so make that an invariant.

Reply:
I agree this is dodgy: If the offsets aren't constant, then we can only give a definitive answer when the `expr.id() == ID_equal && equal`.

Reply:
Does indeed look wrong, I moved the `lhs_offset.is_constant()` into an `INVARIANT` (we shouldn't even get to this point if they aren't constant, as @smowton pointed out).